pkg/crcontrollers/csibmnode/controller.go

```diff
@@ -0,0 +1,328 @@
+		if len(bmNodeCRs.Items[i].Spec.Addresses) > 0 && matchedAddresses == len(bmNodeCRs.Items[i].Spec.Addresses) {
+			bmNode = &bmNodeCRs.Items[i]
+			matchedCRs = append(matchedCRs, bmNode.Name)
+			continue
+		}
+		if matchedAddresses > 0 {
+			ll.Warnf("There is CSIBMNode %s that partially match k8s node %s. CSIBMNode.Spec: %v, k8s node addresses: %v",
+				bmNodeCRs.Items[i].Name, k8sNode.Name, bmNodeCRs.Items[i].Spec, k8sNode.Status.Addresses)
+			return ctrl.Result{}, nil
+		}
+	}
+
+	if len(matchedCRs) > 1 {
+		ll.Warnf("More then one CSIBMNode CR corresponds to the current k8s node (%d). Matched CSIBMNode CRs: %v", len(matchedCRs), matchedCRs)
+		return ctrl.Result{}, nil
+	}
+
+	// create CSIBMNode CR
+	if len(matchedCRs) == 0 {
+		id := uuid.New().String()
+		bmNodeName := namePrefix + id
+		bmNode = bmc.k8sClient.ConstructCSIBMNodeCR(bmNodeName, api.CSIBMNode{
+			UUID:      id,
+			Addresses: bmc.constructAddresses(k8sNode),
+		})
+		if err := bmc.k8sClient.CreateCR(context.Background(), bmNodeName, bmNode); err != nil {
+			ll.Errorf("Unable to create CSIBMNode CR: %v", err)
+			return ctrl.Result{Requeue: true}, err
+		}
+	}
+
+	bmc.cache.put(k8sNode.Name, bmNode.Name)
+	return bmc.updateAnnotation(k8sNode, bmNode.Spec.UUID)
+}
+
+func (bmc *Controller) reconcileForCSIBMNode(bmNode *nodecrd.CSIBMNode) (ctrl.Result, error) {
+	ll := bmc.log.WithFields(logrus.Fields{
+		"method": "reconcileForCSIBMNode",
+		"name":   bmNode.Name,
+	})
+
+	if len(bmNode.Spec.Addresses) == 0 {
+		err := errors.New("addresses are missing for current CSIBMNode instance")
+		ll.Error(err)
+		return ctrl.Result{Requeue: false}, err
+	}
+
+	// get corresponding k8s node name from cache
+	var k8sNode = &coreV1.Node{}
+	if k8sNodeName, ok := bmc.cache.getK8sNodeName(bmNode.Name); ok {
+		if err := bmc.k8sClient.ReadCR(context.Background(), k8sNodeName, k8sNode); err != nil {
+			ll.Errorf("Unable to read k8s node %s: %v", k8sNodeName, err)
+			return ctrl.Result{Requeue: true}, err
+		}
+		return bmc.updateAnnotation(k8sNode, bmNode.Spec.UUID)
+	}
+
+	// search corresponding k8s node name in k8s API
+	k8sNodes := new(coreV1.NodeList)
+	if err := bmc.k8sClient.ReadList(context.Background(), k8sNodes); err != nil {
+		ll.Errorf("Unable to read k8s nodes list: %v", err)
+		return ctrl.Result{Requeue: true}, err
+	}
+
+	matchedNodes := make([]string, 0)
+	for i := range k8sNodes.Items {
+		matchedAddresses := bmc.matchedAddressesCount(bmNode, &k8sNodes.Items[i])
+		if matchedAddresses == len(bmNode.Spec.Addresses) {
+			k8sNode = &k8sNodes.Items[i]
+			matchedNodes = append(matchedNodes, k8sNode.Name)
+			continue
+		}
+		if matchedAddresses > 0 {
+			ll.Warnf("There is k8s node %s that partially match CSIBMNode CR %s. CSIBMNode.Spec: %v, k8s node addresses: %v",
+				k8sNodes.Items[i].Name, bmNode.Name, bmNode.Spec, k8sNodes.Items[i].Status.Addresses)
+			return ctrl.Result{}, nil
+		}
+	}
+
+	switch len(matchedNodes) {
```

Comment:
Maybe we can use single if condition here?
```
	if len(matchedNodes) == 1 {
		bmc.cache.put(k8sNode.Name, bmNode.Name)
		return bmc.updateAnnotation(k8sNode, bmNode.Spec.UUID)
	}
	ll.Warnf("Unable to detect k8s node that corresponds to CSIBMNode %v, matched nodes: %v", bmNode, matchedNodes)
	return ctrl.Result{}, nil
```

Reply:
done